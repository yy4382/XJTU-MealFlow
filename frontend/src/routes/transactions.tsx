import { createFileRoute } from '@tanstack/react-router'
import { useQuery } from '@tanstack/react-query'
import type {
  ColumnDef,
  HeaderContext,
  CellContext,
  Row,
  ColumnFiltersState,
} from '@tanstack/react-table'
import {
  flexRender,
  getCoreRowModel,
  useReactTable,
  getPaginationRowModel,
  getFilteredRowModel,
  getSortedRowModel,
} from '@tanstack/react-table'
import { ArrowUpDown, Download } from 'lucide-react'
import * as React from 'react'

import { fetchAllTransactions, exportCsv } from '../lib/api'
import type { Transaction } from '../lib/types'
import { Button } from '../components/ui/button'
import { Input } from '../components/ui/input'
import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableHeader,
  TableRow,
} from '../components/ui/table'
import {
  Card,
  CardContent,
  CardFooter,
  CardHeader,
} from '../components/ui/card'

export const Route = createFileRoute('/transactions')({
  component: TransactionsPage,
})

const columns: ColumnDef<Transaction>[] = [
  {
    accessorKey: 'time',
    header: ({ column }: HeaderContext<Transaction, unknown>) => {
      return (
        <Button
          variant="ghost"
          onClick={() => column.toggleSorting(column.getIsSorted() === 'asc')}
        >
          Date
          <ArrowUpDown className="ml-2 h-4 w-4" />
        </Button>
      )
    },
    cell: ({ row }: CellContext<Transaction, unknown>) => {
      const timeValue = row.getValue('time')
      const date = new Date(String(timeValue))
      const year = date.getFullYear()
      const month = String(date.getMonth() + 1).padStart(2, '0')
      const day = String(date.getDate()).padStart(2, '0')
      const hours = String(date.getHours()).padStart(2, '0')
      const minutes = String(date.getMinutes()).padStart(2, '0')
      const formattedDate = `${year}-${month}-${day} ${hours}:${minutes}`
      return <div className="font-medium">{formattedDate}</div>
    },
  },
  {
    accessorKey: 'merchant',
    header: 'Merchant',
    cell: ({ row }: CellContext<Transaction, unknown>) => (
      <div>{row.getValue('merchant')}</div>
    ),
  },
  {
    accessorKey: 'amount',
    header: ({ column }: HeaderContext<Transaction, unknown>) => {
      return (
        <Button
          variant="ghost"
          onClick={() => column.toggleSorting(column.getIsSorted() === 'asc')}
          className="text-right w-full justify-end"
        >
          Amount
          <ArrowUpDown className="ml-2 h-4 w-4" />
        </Button>
      )
    },
    cell: ({ row }: CellContext<Transaction, unknown>) => {
      const amount = row.getValue('amount')
      const formatted = amount as number
      return <div className="text-right font-medium">{formatted}</div>
    },
  },
]

function TransactionsPage() {
  const {
    data: transactions,
    isLoading,
    error,
  } = useQuery<Transaction[], Error>({
    queryKey: ['transactions'],
    queryFn: fetchAllTransactions,
  })

  const [columnFilters, setColumnFilters] = React.useState<ColumnFiltersState>(
    [],
  )
  const [isExporting, setIsExporting] = React.useState(false)

  const table = useReactTable<Transaction>({
    data: transactions ?? [],
    columns,
    getCoreRowModel: getCoreRowModel(),
    getPaginationRowModel: getPaginationRowModel(),
    onColumnFiltersChange: setColumnFilters,
    getFilteredRowModel: getFilteredRowModel(),
    getSortedRowModel: getSortedRowModel(),
    state: {
      columnFilters,
    },
    initialState: {
      pagination: {
        pageSize: 15,
      },
      sorting: [
        {
          id: 'time',
          desc: true,
        },
      ],
    },
  })

  const handleExportCsv = async () => {
    try {
      setIsExporting(true)

      // Extract current filters to apply to export
      const merchantFilter = table
        .getColumn('merchant')
        ?.getFilterValue() as string
      // const timeFilter = table.getColumn('time')?.getFilterValue() as string

      const exportParams: any = {}
      if (merchantFilter) {
        exportParams.merchant = merchantFilter
      }

      // For time filter, we might need to parse it depending on how the user enters it
      // For now, we'll export all data matching the merchant filter

      const blob = await exportCsv(exportParams)

      // Create download link
      const url = window.URL.createObjectURL(blob)
      const link = document.createElement('a')
      link.href = url
      link.download = `transactions_export_${new Date().toISOString().split('T')[0]}.csv`
      document.body.appendChild(link)
      link.click()
      document.body.removeChild(link)
      window.URL.revokeObjectURL(url)
    } catch {
      console.error('Export failed:', error)
      alert('Export failed. Please try again.')
    } finally {
      setIsExporting(false)
    }
  }

  if (isLoading) return <div className="p-4">Loading transactions...</div>
  if (error)
    return (
      <div className="p-4 text-red-500">
        Error loading transactions: {error.message}
      </div>
    )

  return (
    <div className="p-4">
      <div className="mb-6">
        <h1 className="text-3xl font-bold tracking-tight mb-2">Transactions</h1>
        <p className="text-muted-foreground">
          View and manage your transaction history.
        </p>
      </div>
      <Card>
        <CardHeader>{/* <CardTitle>Transactions</CardTitle> */}</CardHeader>
        <CardContent>
          <div className="flex items-center justify-between py-4">
            <div className="flex items-center space-x-2">
              <Input
                placeholder="Filter by date..."
                value={table.getColumn('time')?.getFilterValue() as string}
                onChange={(event) =>
                  table.getColumn('time')?.setFilterValue(event.target.value)
                }
                className="max-w-sm"
              />
              <Input
                placeholder="Filter by merchant..."
                value={table.getColumn('merchant')?.getFilterValue() as string}
                onChange={(event) =>
                  table
                    .getColumn('merchant')
                    ?.setFilterValue(event.target.value)
                }
                className="max-w-sm"
              />
            </div>
            <Button
              onClick={handleExportCsv}
              disabled={isExporting}
              variant="outline"
              size="sm"
            >
              <Download className="mr-2 h-4 w-4" />
              {isExporting ? 'Exporting...' : 'Export CSV'}
            </Button>
          </div>
          <div className="rounded-md border">
            <Table>
              <TableHeader>
                {table.getHeaderGroups().map((headerGroup) => (
                  <TableRow key={headerGroup.id}>
                    {headerGroup.headers.map((header) => (
                      <TableHead
                        key={header.id}
                        className={
                          header.column.id === 'amount' ? 'text-right' : ''
                        }
                      >
                        {header.isPlaceholder
                          ? null
                          : flexRender(
                              header.column.columnDef.header,
                              header.getContext(),
                            )}
                      </TableHead>
                    ))}
                  </TableRow>
                ))}
              </TableHeader>
              <TableBody>
                {table.getRowModel().rows.length ? (
                  table.getRowModel().rows.map((row: Row<Transaction>) => (
                    <TableRow
                      key={row.id}
                      data-state={row.getIsSelected() && 'selected'}
                    >
                      {row.getVisibleCells().map((cell) => (
                        <TableCell
                          key={cell.id}
                          className={
                            cell.column.id === 'amount' ? 'text-right' : ''
                          }
                        >
                          {flexRender(
                            cell.column.columnDef.cell,
                            cell.getContext(),
                          )}
                        </TableCell>
                      ))}
                    </TableRow>
                  ))
                ) : (
                  <TableRow>
                    <TableCell
                      colSpan={columns.length}
                      className="h-24 text-center"
                    >
                      No results.
                    </TableCell>
                  </TableRow>
                )}
              </TableBody>
            </Table>
          </div>
        </CardContent>
        <CardFooter>
          <div className="flex items-center justify-end w-full space-x-2 py-4">
            <Button
              variant="outline"
              size="sm"
              onClick={() => table.previousPage()}
              disabled={!table.getCanPreviousPage()}
            >
              Previous
            </Button>
            <Button
              variant="outline"
              size="sm"
              onClick={() => table.nextPage()}
              disabled={!table.getCanNextPage()}
            >
              Next
            </Button>
          </div>
        </CardFooter>
      </Card>
    </div>
  )
}
