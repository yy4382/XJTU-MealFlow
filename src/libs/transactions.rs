use std::{path::PathBuf, rc::Rc};

use chrono::{DateTime, Local};
use color_eyre::eyre::{Result, bail};
use rusqlite::{Connection, params};

#[derive(Debug, Clone)]
pub struct Transaction {
    pub id: u64,
    pub time: DateTime<Local>,
    pub amount: f64,
    pub merchant: String,
}

#[derive(Debug, Clone)]
pub struct TransactionManager {
    conn: Rc<Connection>,
}

impl TransactionManager {
    pub fn new(db_path: Option<PathBuf>) -> Result<Self> {
        let conn = match db_path {
            Some(db_path) => {
                std::fs::create_dir_all(db_path.parent().unwrap())?;
                Connection::open(db_path)?
            }
            None => Connection::open_in_memory()?,
        };

        Ok(TransactionManager {
            conn: Rc::new(conn),
        })
    }

    pub fn init_db(&self) -> Result<(), rusqlite::Error> {
        self.conn.execute(
            "CREATE TABLE IF NOT EXISTS transactions (
                id INTEGER PRIMARY KEY,
                time TEXT NOT NULL,
                amount REAL NOT NULL,
                merchant TEXT NOT NULL
            )",
            [],
        )?;
        self.conn.execute(
            "CREATE TRIGGER IF NOT EXISTS prevent_transaction_conflict 
                BEFORE INSERT ON transactions
                FOR EACH ROW
                BEGIN
                    SELECT CASE
                    WHEN EXISTS (
                        SELECT 1 FROM transactions 
                        WHERE id = NEW.id 
                        AND time = NEW.time 
                        AND amount = NEW.amount 
                        AND merchant = NEW.merchant
                    ) THEN
                        RAISE(IGNORE)  -- 完全相同的记录则静默跳过
                    WHEN EXISTS (
                        SELECT 1 FROM transactions 
                        WHERE id = NEW.id
                    ) THEN
                        RAISE(ABORT, 'Conflict: Existing transaction with different data')  -- ID存在但数据不同时终止
                    END;
                END;",
            [],
        )?;
        self.conn.execute(
            "CREATE TABLE IF NOT EXISTS cookies (
            account TEXT PRIMARY KEY,
            cookie TEXT NOT NULL
        )",
            [],
        )?;
        Ok(())
    }

    pub fn insert(&self, transactions: &Vec<Transaction>) -> Result<(), rusqlite::Error> {
        // insert at once
        let mut stmt = self
            .conn
            .prepare("INSERT INTO transactions (id, time, amount, merchant) VALUES (?, ?, ?, ?)")?;

        for transaction in transactions {
            stmt.execute(params![
                transaction.id,
                transaction.time.to_rfc3339(),
                transaction.amount,
                transaction.merchant
            ])?;
        }
        Ok(())
    }

    pub fn fetch_all(&self) -> Result<Vec<Transaction>, rusqlite::Error> {
        let mut stmt = self
            .conn
            .prepare("SELECT id, time, amount, merchant FROM transactions")?;
        let transactions = stmt.query_map([], |row| {
            let time_str: String = row.get(1)?;
            let time = chrono::DateTime::parse_from_rfc3339(&time_str)
                .map_err(|e| rusqlite::Error::InvalidParameterName(e.to_string()))?
                .with_timezone(&Local);

            Ok(Transaction {
                id: row.get(0)?,
                time,
                amount: row.get(2)?,
                merchant: row.get(3)?,
            })
        })?;

        let mut result = Vec::new();
        for transaction in transactions {
            result.push(transaction?);
        }
        Ok(result)
    }

    pub fn fetch_count(&self) -> Result<u64, rusqlite::Error> {
        let mut stmt = self.conn.prepare("SELECT COUNT(*) FROM transactions")?;
        let count = stmt.query_map([], |row| row.get(0))?;
        let mut result = 0;
        for c in count {
            result = c?;
        }
        Ok(result)
    }

    #[allow(dead_code)]
    pub fn clear_db(&self) -> Result<(), rusqlite::Error> {
        self.conn.execute("DELETE FROM transactions", [])?;
        Ok(())
    }

    /// Update account(optional) and cookie in cookies table
    ///
    /// If there is already a record, update it. Otherwise, insert a new record.
    /// There should always be only one records in cookies table
    pub fn update_account(&self, account: &str) -> Result<()> {
        // Check if there are existing records
        let existing = self
            .conn
            .query_row("SELECT cookie FROM cookies LIMIT 1", [], |row| {
                row.get::<_, String>(0)
            });

        // Determine account value to use
        let cookie = existing.unwrap_or_default();

        // Replace the record
        self.conn.execute("DELETE FROM cookies", [])?;
        let mut stmt = self
            .conn
            .prepare("INSERT INTO cookies (account, cookie) VALUES (?, ?)")?;
        stmt.execute(params![account, cookie])?;
        Ok(())
    }

    pub fn update_cookie(&self, cookie: &str) -> Result<()> {
        // Check if there are existing records
        let existing = self
            .conn
            .query_row("SELECT account FROM cookies LIMIT 1", [], |row| {
                row.get::<_, String>(0)
            });

        // Determine account value to use
        let account = existing.unwrap_or_default();

        // Replace the record
        self.conn.execute("DELETE FROM cookies", [])?;
        let mut stmt = self
            .conn
            .prepare("INSERT INTO cookies (account, cookie) VALUES (?, ?)")?;
        stmt.execute(params![account, cookie])?;
        Ok(())
    }

    pub fn get_account_cookie(&self) -> Result<(String, String)> {
        let mut stmt = self.conn.prepare("SELECT account, cookie FROM cookies")?;
        let mut rows = stmt.query([])?;
        let row = rows.next()?;
        match row {
            Some(row) => {
                let account = row.get(0)?;
                let cookie = row.get(1)?;
                Ok((account, cookie))
            }
            None => bail!("No account and cookie found"),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_transaction_manager() {
        let manager = TransactionManager::new(None).unwrap();

        manager.init_db().unwrap();

        manager.clear_db().unwrap();

        let transactions = vec![
            Transaction {
                id: 1,
                time: Local::now(),
                amount: -100.0,
                merchant: "Amazon".to_string(),
            },
            Transaction {
                id: 2,
                time: Local::now(),
                amount: -200.0,
                merchant: "Google".to_string(),
            },
        ];

        manager.insert(&transactions).unwrap();

        let fetched = manager.fetch_all().unwrap();
        assert_eq!(fetched.len(), 2);
        assert_eq!(fetched[0].id, 1);
        assert_eq!(fetched[0].amount, -100.0);
        assert_eq!(fetched[0].merchant, "Amazon");

        assert_eq!(fetched[1].id, 2);
        assert_eq!(fetched[1].amount, -200.0);
        assert_eq!(fetched[1].merchant, "Google");
    }

    #[test]
    fn test_account_cookie() {
        let manager = TransactionManager::new(None).unwrap();
        manager.init_db().unwrap();

        manager.update_account("test_account").unwrap();
        let (account, cookie) = manager.get_account_cookie().unwrap();
        assert_eq!(account, "test_account");
        assert_eq!(cookie, "");

        manager.update_cookie("test_cookie").unwrap();
        let (account, cookie) = manager.get_account_cookie().unwrap();
        assert_eq!(account, "test_account");
        assert_eq!(cookie, "test_cookie");

        manager.update_account("test_account2").unwrap();
        let (account, cookie) = manager.get_account_cookie().unwrap();
        assert_eq!(account, "test_account2");
        assert_eq!(cookie, "test_cookie");
    }

    #[test]
    fn test_fetch_count() {
        let manager = TransactionManager::new(None).unwrap();
        manager.init_db().unwrap();
        manager.clear_db().unwrap();

        // Initially should have zero transactions
        let count = manager.fetch_count().unwrap();
        assert_eq!(count, 0);

        // Add some transactions
        let transactions = vec![
            Transaction {
                id: 1,
                time: Local::now(),
                amount: -100.0,
                merchant: "Amazon".to_string(),
            },
            Transaction {
                id: 2,
                time: Local::now(),
                amount: -200.0,
                merchant: "Google".to_string(),
            },
        ];

        manager.insert(&transactions).unwrap();

        // Should now have 2 transactions
        let count = manager.fetch_count().unwrap();
        assert_eq!(count, 2);

        // Add more transactions
        let more_transactions = vec![Transaction {
            id: 3,
            time: Local::now(),
            amount: -300.0,
            merchant: "Apple".to_string(),
        }];

        manager.insert(&more_transactions).unwrap();

        // Should now have 3 transactions
        let count = manager.fetch_count().unwrap();
        assert_eq!(count, 3);

        // Clear the database
        manager.clear_db().unwrap();

        // Should now have 0 transactions again
        let count = manager.fetch_count().unwrap();
        assert_eq!(count, 0);
    }
}
