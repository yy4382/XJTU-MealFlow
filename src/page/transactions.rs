use crate::{RootState, actions::Action, tui::Event};

use super::Page;
use color_eyre::eyre::Result;
use crossterm::event::KeyCode;
use ratatui::{
    Frame,
    layout::{Alignment, Constraint, Layout},
    style::{Color, Style},
    text::Text,
    widgets::{Block, BorderType, Borders, Paragraph},
};

#[derive(Default, Clone)]
pub struct Transactions {
    transactions: Vec<crate::libs::transactions::Transaction>,
}

#[derive(Clone)]
pub enum TransactionAction {
    LoadTransactions,
}

impl Page for Transactions {
    fn render(&self, frame: &mut Frame, _app: &RootState) {
        let area = frame.area();

        let vertical = &Layout::vertical([Constraint::Fill(1), Constraint::Length(3)]);
        let rects = vertical.split(area);

        frame.render_widget(
            {
                let transactions_str = self
                    .transactions
                    .iter()
                    .fold(String::new(), |acc, x| acc + &format!("{:?}\n", x));
                Paragraph::new(transactions_str)
                    .block(
                        Block::default()
                            .title("Transactions")
                            .title_alignment(Alignment::Center)
                            .borders(Borders::ALL)
                            .border_type(BorderType::Rounded),
                    )
                    .style(Style::default().fg(Color::Cyan))
                    .alignment(Alignment::Center)
            },
            rects[0],
        );

        frame.render_widget(
            Paragraph::new(Text::raw("r: Refresh l: Load")).block(
                Block::default()
                    .borders(Borders::ALL)
                    .border_type(BorderType::Rounded),
            ),
            rects[1],
        );
    }
    fn handle_events(&self, app: &RootState, event: Event) -> Result<()> {
        match event {
            Event::Key(key) => match (key.modifiers, key.code) {
                // navigate to fetch page
                (_, KeyCode::Char('r')) => app.send_action(Action::NavigateTo(
                    crate::actions::NaviTarget::Fetch(crate::page::fetch::Fetch::default()),
                )),
                (_, KeyCode::Char('l')) => {
                    app.send_action(Action::Transaction(TransactionAction::LoadTransactions))
                }
                _ => (),
            },
            _ => (),
        };
        Ok(())
    }

    fn update(&mut self, root_state: &RootState, action: Action) {
        if let Action::Transaction(action) = action {
            match action {
                TransactionAction::LoadTransactions => {
                    self.transactions = root_state.manager.fetch_all().unwrap();
                    root_state.send_action(Action::Render);
                }
            }
        }
    }

    fn get_name(&self) -> String {
        "Transactions".to_string()
    }

    fn init(&mut self, app: &mut RootState) {
        app.send_action(Action::Transaction(TransactionAction::LoadTransactions))
    }
}
