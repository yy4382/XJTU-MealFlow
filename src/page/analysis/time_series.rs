use std::collections::HashMap;

use chrono::Datelike;
use ratatui::{
    Frame,
    style::{Style, Stylize as _, palette::tailwind},
    symbols,
    text::Line,
    widgets::{Bar, BarChart, BarGroup, Block, Padding, Paragraph},
};
use tracing::info;

use crate::libs::transactions::Transaction;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
struct YearMonth {
    year: u16,
    month: u16,
}
impl YearMonth {
    fn new(year: u16, month: u16) -> Self {
        Self { year, month }
    }
    // fn relative_to(&self, other: &Self) -> f64 {
    //     let year_diff = self.year as f64 - other.year as f64;
    //     let month_diff = self.month as f64 - other.month as f64;
    //     year_diff * 12.0 + month_diff
    // }
    // fn relative_to_default(&self) -> f64 {
    //     let default = YearMonth::new(2015, 1);
    //     self.relative_to(&default)
    // }
    fn next(&self) -> Self {
        if self.month == 12 {
            YearMonth::new(self.year + 1, 1)
        } else {
            YearMonth::new(self.year, self.month + 1)
        }
    }
}
impl std::fmt::Display for YearMonth {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:02}-{:02}", self.year, self.month)
    }
}

#[derive(Debug, Default, Clone)]
pub(super) struct TimeSeriesData {
    data: Vec<(YearMonth, f64)>,
}

impl TimeSeriesData {
    pub(super) fn new(data: &[Transaction]) -> Self {
        let mut processed_data: Vec<(YearMonth, f64)> = data
            .iter()
            .fold(HashMap::new(), |mut acc, entry| {
                let year = entry.time.year() as u16;
                let month = entry.time.month() as u16;
                match acc.get(&(year, month)) {
                    Some(v) => acc.insert((year, month), *v + entry.amount.abs()),
                    None => acc.insert((year, month), entry.amount.abs()),
                };
                acc
            })
            .iter()
            .map(|((year, month), amount)| (YearMonth::new(*year, *month), *amount))
            .collect();

        processed_data.sort_by(|a, b| a.0.cmp(&b.0));

        let processed_data = processed_data
            .into_iter()
            .fold(Vec::new(), |mut acc, entry| {
                let mut last_ym = acc.last().unwrap_or(&entry).0;
                while last_ym.next() < entry.0 {
                    info!("Missing data for {:?}, {:?}", last_ym, entry.0);
                    acc.push((last_ym, 0.0));
                    last_ym = last_ym.next();
                }
                acc.push(entry);
                acc
            });

        info!("{:?}", processed_data);

        Self {
            data: processed_data,
        }
    }
    pub(super) fn render(
        &self,
        area: ratatui::prelude::Rect,
        frame: &mut Frame,
        color: tailwind::Palette,
    ) {
        let block = Block::bordered()
            .border_set(symbols::border::PROPORTIONAL_TALL)
            .border_style(color.c600)
            .padding(Padding::horizontal(1));

        if self.data.len() == 0 {
            frame.render_widget(
                Paragraph::new("No data available yet").block(block.clone()),
                area,
            );
            return;
        }

        let bars = (area.width as u64 - 6) / 8;

        let style = Style::default().fg(color.c300);
        let bars: Vec<Bar> = self.data[self.data.len().saturating_sub(bars as usize)..]
            .iter()
            .map(|(ym, value)| {
                Bar::default()
                    .value(value.round() as u64)
                    .label(Line::from(ym.to_string()))
                    .style(style)
                    .value_style(style.reversed())
            })
            .collect();

        let bar_chart = BarChart::default()
            .block(block)
            .data(BarGroup::default().bars(&bars))
            .bar_width(7)
            .bar_gap(1)
            .bar_style(style);

        frame.render_widget(bar_chart, area);
    }
}

#[cfg(test)]
mod test {
    use insta::assert_snapshot;
    use ratatui::backend::TestBackend;

    use super::*;

    #[test]
    fn test_empty_render() {
        let mut terminal = ratatui::Terminal::new(TestBackend::new(80, 20)).unwrap();
        let data = TimeSeriesData::default();
        terminal
            .draw(|f| data.render(f.area(), f, tailwind::BLUE))
            .unwrap();
        assert_snapshot!(terminal.backend())
    }
}
