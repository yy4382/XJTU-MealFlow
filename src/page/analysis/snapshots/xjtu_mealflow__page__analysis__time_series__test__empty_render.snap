---
source: src/page/analysis/time_series.rs
expression: terminal.backend()
---
"█▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀█"
"█ No data available yet                                                        █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█                                                                              █"
"█▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄█"
